narwhal/primary/src/tests/core_tests.rs

```diff
@@ -1028,6 +1028,128 @@ async fn recover_core_expecting_header_of_previous_round() {
     }
 }
 
+#[tokio::test(flavor = "current_thread", start_paused = true)]
+async fn recover_core_latest_headers() {
+    let fixture = CommitteeFixture::builder().randomize_ports(true).build();
+    let committee = fixture.committee();
+    let worker_cache = fixture.shared_worker_cache();
+    let primary = fixture.authorities().last().unwrap();
+    let name = primary.public_key();
+    let signature_service = SignatureService::new(primary.keypair().copy());
+
+    let (_tx_reconfigure, rx_reconfigure) =
+        watch::channel(ReconfigureNotification::NewEpoch(committee.clone()));
+    let (tx_sync_headers, _rx_sync_headers) = test_utils::test_channel!(1);
+    let (tx_certificate_waiter, _rx_certificate_waiter) = test_utils::test_channel!(1);
+    let (_tx_primary_messages, rx_primary_messages) = test_utils::test_channel!(3);
+    let (_tx_headers_loopback, rx_headers_loopback) = test_utils::test_channel!(1);
+    let (_tx_certificates_loopback, rx_certificates_loopback) = test_utils::test_channel!(1);
+    let (_tx_headers, rx_headers) = test_utils::test_channel!(1);
+    let (tx_consensus, _rx_consensus) = test_utils::test_channel!(3);
+    let (tx_parents, _rx_parents) = test_utils::test_channel!(3);
+    let (_tx_consensus_round_updates, rx_consensus_round_updates) = watch::channel(0u64);
+
+    // Create test stores.
+    let (header_store, certificates_store, payload_store) = create_db_stores();
+
+    // Make a synchronizer for the core.
+    let synchronizer = Synchronizer::new(
+        name.clone(),
+        &committee,
+        certificates_store.clone(),
+        payload_store.clone(),
+        /* tx_header_waiter */ tx_sync_headers.clone(),
+        tx_certificate_waiter.clone(),
+        None,
+    );
+
+    let metrics = Arc::new(PrimaryMetrics::new(&Registry::new()));
+
+    let own_address = network::multiaddr_to_address(&committee.primary(&name).unwrap()).unwrap();
+    println!("New primary added: {:?}", own_address);
+    let network = anemo::Network::bind(own_address)
+        .server_name("narwhal")
+        .private_key(primary.network_keypair().copy().private().0.to_bytes())
+        .start(anemo::Router::new())
+        .unwrap();
+
+    let mut primary_networks = Vec::new();
+    let fixture_authorities = fixture.authorities().filter(|a| a.public_key() != name);
+    for primary in fixture_authorities {
+        let address = committee.primary(&primary.public_key()).unwrap();
+
+        let mut mock_server = MockPrimaryToPrimary::new();
+        mock_server
+            .expect_get_latest_header()
+            .returning(move |_request| {
+                let builder = types::HeaderBuilder::default();
+                let fixture = CommitteeFixture::builder().randomize_ports(true).build();
```

Comment:
can you avoid building CommitteeFixture a second time in the mock handler? I think you move your "let primary" line on 1087 up to 1080, hopefully it will be possible for it to be moved into the closure and the rest will work as-is?

Reply:
It wasn't possible because the fixture can't be guaranteed to live long enough

Reply:
The fixture doesn't need to be alive, all the mock handler needs is the Header

This works:

```rust
        let address = committee.primary(&primary.public_key()).unwrap();
        let builder = types::HeaderBuilder::default();
        let header = builder
            .author(primary.public_key())
            .round(1)
            .epoch(0)
            .parents([CertificateDigest::default()].iter().cloned().collect())
            .with_payload_batch(fixture_batch_with_transactions(10), 0)
            .build(primary.keypair())
            .unwrap();

        let mut mock_server = MockPrimaryToPrimary::new();
        mock_server
            .expect_get_latest_header()
            .returning(move |_request| {
                let header = header.clone();
                Ok(anemo::Response::new(LatestHeaderResponse {
                    header: Some(header),
                }))
            });```